### PR TITLE
 More precise digamma

### LINKED ISSUE #6190: [PyTorch]: bad digamma float32 accuracy
PyTorch master@92a0f78

Our digamma single-precision floating point accuracy is bad near the poles. This is not caught by our tests because we only test on a few values:

https://github.com/pytorch/pytorch/blob/92a0f7835e7e2f6342fe2fd32a25299ea86aed31/test/test_torch.py#L272-L274

```python
input = torch.tensor([-1.99999994], dtype=torch.float32)
fp32 = input.digamma()
fp64 = input.double().digamma()
print(((fp32.double() - fp64) / fp64).item())  # 0.24012388522631392
```

This is a relative error of **24%**

We also are returning real numbers at the poles (for both float32 and float64) when we should be returning inf.

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #6190.

This is a rebase of #3955 with some tweaks for better performance around
poles. The code is ported over from cephes with permission.

By itself, the cephes code returns inf for the poles.

For better performance around the poles with float32, one intermediate
step is always computed with double precision, regardless of dtype.
This step does `PI / tan(PI * input)`. This is necessary because small (1e-6)
rounding errors for the inputs to tan have strong effects on the output
(ie, the derivative of tan is very large at some points).

cc @colesbury @apaszke

### CODE REVIEW

**Comment by SsnL** [comment]
Claiming to remind me to review later. If anyone feels like reviewing, please feel free to do so. :)

File: aten/src/TH/THMath.h [hunk 1]
```diff
@@ -129,34 +129,123 @@ Date:  February 1996
...
-  if (x < 0.5) {
-    result -= M_PI / tan(M_PI * x);
-    x = 1 - x;
+  for (size_t i = 0; i <= len; i++) {
+    result = result * x + A[i];
+  }
+  return result;
+}
+
+static inline float TH_polevlf(float x, float *A, size_t len) {
```
Comment (apaszke) [inline aten/src/TH/THMath.h]: Can we use templates to deduplicate those?

File: aten/src/THC/THCTensorMathPointwise.cuh [hunk 2]
```diff
@@ -818,25 +818,73 @@ struct TensorBitXorOp {
...
+       7.57575757575757575758E-3,
+      -4.16666666666666666667E-3,
+       3.96825396825396825397E-3,
+      -8.33333333333333333333E-3,
+       8.33333333333333333333E-2,
+    };
+
     compute_type x = ScalarConvert<real, compute_type>::to(*in);
+    if (x == 0) {
+      *out = ScalarConvert<float, real>::to(INFINITY);
```
Comment (apaszke) [inline aten/src/THC/THCTensorMathPointwise.cuh]: Let's use `scalar_cast`. It's much more readable

**Review by apaszke — APPROVED** [review]
Haven't reviewed the math (I assume it just comes from Cephes). Two minor things, but LGTM.

File: aten/src/TH/THMath.h (hunk 1 above)
Comment (zou3519) [reply aten/src/TH/THMath.h]: I think this is a header file for a C file so no, I don't think we can use templates here :(

Comment (SsnL) [reply aten/src/TH/THMath.h]: I think we can make it .cpp, right?

Comment (zou3519) [reply aten/src/TH/THMath.h]: I think I'd have to convert THTensorMath.c into a cpp file or make a new one for polevl. I'll leave this for now; there's a lot of repeated code in THMath.h that can benefit from a future cleanup like this.
